File: core/runtime.py
"""运行时判定与目录约定。"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from enum import Enum
from pathlib import Path


APP_DIR_NAME = "ATRI-Chat"
PORTABLE_MARKER = "portable.mode"
BINARIES_DIR_NAME = "binaries"
RESOURCES_DIR_NAME = "resources"
RUNTIME_MODE_ENV = "ATRI_RUNTIME_MODE"
APP_ENV_ENV = "ATRI_APP_ENV"
APP_ROOT_ENV = "ATRI_APP_ROOT"
DATA_ROOT_ENV = "ATRI_DATA_ROOT"
LOGS_ROOT_ENV = "ATRI_LOGS_ROOT"


def normalize_windows_path(path: str | Path) -> Path:
    """移除 Windows verbatim 路径前缀，避免下游 SQLite/日志库兼容性问题。"""
    raw = str(path)
    if os.name == "nt":
        if raw.startswith("\\\\?\\UNC\\"):
            return Path("\\\\" + raw[8:])
        if raw.startswith("\\\\?\\"):
            return Path(raw[4:])
    return Path(raw)


def build_runtime_layout(
    app_env: AppEnv,
    mode: RuntimeMode,
    app_root: str | Path,
    data_root: str | Path,
    logs_root: str | Path,
) -> "RuntimeLayout":
    """构造统一的运行时布局，并在入口处完成路径规整。"""
    return RuntimeLayout(
        app_env=app_env,
        mode=mode,
        app_root=normalize_windows_path(app_root).resolve(),
        data_root=normalize_windows_path(data_root).resolve(),
        logs_root=normalize_windows_path(logs_root).resolve(),
    )


class AppEnv(str, Enum):
    """应用环境。"""

    DEVELOPMENT = "development"
    PRODUCTION = "production"


class RuntimeMode(str, Enum):
    """桌面端运行模式。"""

    DEVELOPMENT = "development"
    INSTALLED = "installed"
    PORTABLE = "portable"


@dataclass(frozen=True)
class RuntimeLayout:
    """运行时判定结果。"""

    app_env: AppEnv
    mode: RuntimeMode
    app_root: Path
    data_root: Path
    logs_root: Path


def get_project_root() -> Path:
    """返回仓库根目录。"""
    return normalize_windows_path(Path(__file__).resolve().parent.parent)


def get_local_data_home() -> Path:
    """返回本地用户数据根目录。"""
    local_appdata = os.environ.get("LOCALAPPDATA")
    if local_appdata:
        return normalize_windows_path(local_appdata).resolve() / APP_DIR_NAME
    return get_project_root()


def resolve_data_root(runtime_mode: RuntimeMode, app_root: Path) -> Path:
    """根据运行模式解析数据根目录。"""
    if runtime_mode == RuntimeMode.INSTALLED:
        return get_local_data_home() / "data"
    return app_root / "data"


def resolve_logs_root(runtime_mode: RuntimeMode, app_root: Path) -> Path:
    """根据运行模式解析日志根目录。"""
    if runtime_mode == RuntimeMode.INSTALLED:
        return get_local_data_home() / "logs"
    return app_root / "data" / "logs"


def resolve_runtime_layout() -> RuntimeLayout:
    """根据实际运行情况判定目录策略。"""
    env_layout = resolve_runtime_layout_from_env()
    if env_layout is not None:
        return env_layout

    if not getattr(sys, "frozen", False):
        app_root = get_project_root().resolve()
        mode = RuntimeMode.DEVELOPMENT
        return build_runtime_layout(
            app_env=AppEnv.DEVELOPMENT,
            mode=mode,
            app_root=app_root,
            data_root=resolve_data_root(mode, app_root),
            logs_root=resolve_logs_root(mode, app_root),
        )

    executable = normalize_windows_path(sys.executable).resolve()
    app_root = resolve_packaged_app_root(executable)
    mode = detect_packaged_mode(app_root)
    return build_runtime_layout(
        app_env=AppEnv.PRODUCTION,
        mode=mode,
        app_root=app_root,
        data_root=resolve_data_root(mode, app_root),
        logs_root=resolve_logs_root(mode, app_root),
    )


def resolve_runtime_layout_from_env() -> RuntimeLayout | None:
    """优先使用桌面壳注入的运行时布局，确保前后端共享同一套目录策略。"""
    runtime_mode = os.environ.get(RUNTIME_MODE_ENV)
    app_env = os.environ.get(APP_ENV_ENV)
    app_root = os.environ.get(APP_ROOT_ENV)
    data_root = os.environ.get(DATA_ROOT_ENV)
    logs_root = os.environ.get(LOGS_ROOT_ENV)

    if not all([runtime_mode, app_env, app_root, data_root, logs_root]):
        return None

    return build_runtime_layout(
        app_env=AppEnv(app_env),
        mode=RuntimeMode(runtime_mode),
        app_root=app_root,
        data_root=data_root,
        logs_root=logs_root,
    )


def resolve_packaged_app_root(executable: Path) -> Path:
    """根据打包产物位置反推出应用根目录。"""
    exe_dir = executable.parent
    if exe_dir.name.lower() != BINARIES_DIR_NAME:
        return exe_dir

    parent = exe_dir.parent
    if parent.name.lower() == RESOURCES_DIR_NAME and parent.parent:
        return parent.parent
    return parent


def detect_packaged_mode(app_root: Path) -> RuntimeMode:
    """根据打包目录结构判断安装版或便携版。"""
    if (app_root / PORTABLE_MARKER).exists():
        return RuntimeMode.PORTABLE
    return RuntimeMode.INSTALLED

File: frontend/src-tauri/src/runtime.rs
use std::env;
use std::fs;
use std::io;
use std::ffi::OsString;
use std::path::{Path, PathBuf};

use tauri::Manager;

const APP_DIR_NAME: &str = "ATRI-Chat";
const PORTABLE_MARKER: &str = "portable.mode";
const BINARIES_DIR_NAME: &str = "binaries";
const RESOURCES_DIR_NAME: &str = "resources";
pub const BACKEND_PORT_ENV: &str = "ATRI_BACKEND_PORT";
pub const RUNTIME_MODE_ENV: &str = "ATRI_RUNTIME_MODE";
pub const APP_ENV_ENV: &str = "ATRI_APP_ENV";
pub const APP_ROOT_ENV: &str = "ATRI_APP_ROOT";
pub const DATA_ROOT_ENV: &str = "ATRI_DATA_ROOT";
pub const LOGS_ROOT_ENV: &str = "ATRI_LOGS_ROOT";

#[derive(Clone, Copy, Debug)]
pub enum RuntimeMode {
    Development,
    Installed,
    Portable,
}

impl RuntimeMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Development => "development",
            Self::Installed => "installed",
            Self::Portable => "portable",
        }
    }

    pub fn app_env(&self) -> &'static str {
        match self {
            Self::Development => "development",
            Self::Installed | Self::Portable => "production",
        }
    }
}

#[derive(Clone, Debug)]
pub struct RuntimeLayout {
    pub mode: RuntimeMode,
    pub app_root: PathBuf,
    pub data_root: PathBuf,
    pub logs_root: PathBuf,
}

#[derive(Clone, Debug)]
pub enum BackendLaunch {
    UvProject {
        cwd: PathBuf,
        program: &'static str,
        args: Vec<&'static str>,
    },
    Sidecar(PathBuf),
}

impl RuntimeLayout {
    pub fn new(mode: RuntimeMode, app_root: PathBuf, data_root: PathBuf, logs_root: PathBuf) -> Self {
        Self {
            mode,
            app_root: normalize_windows_path(app_root),
            data_root: normalize_windows_path(data_root),
            logs_root: normalize_windows_path(logs_root),
        }
    }

    pub fn ensure_directories(&self) -> io::Result<()> {
        fs::create_dir_all(&self.data_root)?;
        fs::create_dir_all(&self.logs_root)?;
        Ok(())
    }

    pub fn backend_environment(&self, port: u16) -> Vec<(&'static str, OsString)> {
        vec![
            (BACKEND_PORT_ENV, port.to_string().into()),
            (RUNTIME_MODE_ENV, self.mode.as_str().into()),
            (APP_ENV_ENV, self.mode.app_env().into()),
            (APP_ROOT_ENV, self.app_root.clone().into_os_string()),
            (DATA_ROOT_ENV, self.data_root.clone().into_os_string()),
            (LOGS_ROOT_ENV, self.logs_root.clone().into_os_string()),
        ]
    }
}

fn normalize_windows_path(path: PathBuf) -> PathBuf {
    #[cfg(windows)]
    {
        let raw = path.to_string_lossy();
        if let Some(stripped) = raw.strip_prefix(r"\\?\UNC\") {
            return PathBuf::from(format!(r"\\{}", stripped));
        }
        if let Some(stripped) = raw.strip_prefix(r"\\?\") {
            return PathBuf::from(stripped);
        }
    }

    path
}

pub fn resolve_runtime_layout() -> RuntimeLayout {
    if cfg!(debug_assertions) {
        let app_root = development_app_root();
        return RuntimeLayout::new(
            RuntimeMode::Development,
            app_root.clone(),
            app_root.join("data"),
            app_root.join("data").join("logs"),
        );
    }

    let app_root = packaged_app_root();
    let mode = detect_packaged_mode(&app_root);
    let local_data_root = local_appdata_root()
        .unwrap_or_else(|| app_root.clone())
        .join(APP_DIR_NAME);

    match mode {
        RuntimeMode::Installed => RuntimeLayout::new(
            mode,
            app_root,
            local_data_root.join("data"),
            local_data_root.join("logs"),
        ),
        RuntimeMode::Portable | RuntimeMode::Development => RuntimeLayout::new(
            mode,
            app_root.clone(),
            app_root.join("data"),
            app_root.join("data").join("logs"),
        ),
    }
}

pub fn resolve_backend_path(app: &tauri::AppHandle) -> Option<PathBuf> {
    let backend_name = backend_binary_name();

    if cfg!(debug_assertions) {
        return Some(PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("binaries").join(backend_name));
    }

    let resource_dir = app.path().resource_dir().ok()?;
    Some(resource_dir.join("binaries").join(backend_name))
}

pub fn resolve_backend_launch(app: &tauri::AppHandle, layout: &RuntimeLayout) -> Option<BackendLaunch> {
    if cfg!(debug_assertions) {
        return Some(BackendLaunch::UvProject {
            cwd: layout.app_root.clone(),
            program: "uv",
            args: vec!["run", "python", "main.py"],
        });
    }

    resolve_backend_path(app).map(BackendLaunch::Sidecar)
}

fn detect_packaged_mode(app_root: &Path) -> RuntimeMode {
    if app_root.join(PORTABLE_MARKER).exists() {
        RuntimeMode::Portable
    } else {
        RuntimeMode::Installed
    }
}

fn development_app_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .canonicalize()
        .unwrap_or_else(|_| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..").join(".."))
}

fn packaged_app_root() -> PathBuf {
    env::current_exe()
        .ok()
        .map(|path| resolve_app_root_from_executable(&path))
        .unwrap_or_else(|| env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
}

fn resolve_app_root_from_executable(executable: &Path) -> PathBuf {
    let exe_dir = executable
        .parent()
        .map(|dir| dir.to_path_buf())
        .unwrap_or_else(|| PathBuf::from("."));

    if exe_dir.file_name().and_then(|name| name.to_str()) != Some(BINARIES_DIR_NAME) {
        return exe_dir;
    }

    let parent = exe_dir.parent().map(|dir| dir.to_path_buf()).unwrap_or(exe_dir);
    if parent.file_name().and_then(|name| name.to_str()) == Some(RESOURCES_DIR_NAME) {
        return parent.parent().map(|dir| dir.to_path_buf()).unwrap_or(parent);
    }

    parent
}

fn local_appdata_root() -> Option<PathBuf> {
    env::var_os("LOCALAPPDATA").map(PathBuf::from)
}

fn backend_binary_name() -> &'static str {
    if cfg!(windows) {
        "atri-backend-x86_64-pc-windows-msvc.exe"
    } else if cfg!(target_os = "macos") {
        if cfg!(target_arch = "aarch64") {
            "atri-backend-aarch64-apple-darwin"
        } else {
            "atri-backend-x86_64-apple-darwin"
        }
    } else {
        "atri-backend-x86_64-unknown-linux-gnu"
    }
}
